Improve secret hostname hint

jerolimov opened the pull request:
**Fixes**: 
https://issues.redhat.com/browse/ODC-4906

**Analysis / Root cause**: 
When starting a pipeline there is an option to add a `Secret` to fetch data from a private git repository. The hint here suggest a server name like "https://github.com", but the pipeline/secret combination works only when the server name is something like "github.com".

**Solution Description**: 
Update the hint from static `The base server url (e.g. https://github.com)` to a dynamic version which is now depending on the selected auth mechanism:

For docker config json: `The base server url (e.g. https://quay.io/)` (new example server)
For http basic auth: `The base server url (e.g. https://github.com)` (unchanged)
For SSH: `Server hostname without schema or path (e.g. github.com)`


Edit: After getting some feedback in this PR, this PR now also changes the order of the input fields to show depending fields below the more generic options.

It also hides some useless "Access To" and "Authentication Type" combinations:

* For "Access To: Git" we only show the auth types "Basic Authentication" and "SSH Key" and dropped the option "Image Registry Credentials"
* For "Access To: Image" we only show the auth types "Basic Authentication" and "Image Registry Credentials" and dropeed the option "SSH Key".

See https://github.com/openshift/console/pull/6873#issuecomment-717912374 and https://github.com/openshift/console/pull/6873#issuecomment-718436148

**Screen shots / Gifs for design review**: 
@openshift/team-devconsole-ux

Before:
![old-layout](https://user-images.githubusercontent.com/139310/98481378-1141d600-21fa-11eb-9796-9dcf45ad34c9.png)

After:
![updated-pr](https://user-images.githubusercontent.com/139310/98699544-8dadf380-2377-11eb-9ca0-0fc77870107b.png)

And some more detail screens:
<details>
<summary>Access to options in now above the auth type</summary>
<img src="https://user-images.githubusercontent.com/139310/98481321-c1fba580-21f9-11eb-8853-8731f6b94175.png"/>
</details>
<details>
<summary>Docker registry auth type options:</summary>
<img src="https://user-images.githubusercontent.com/139310/98481324-c45dff80-21f9-11eb-9dc2-94ba6bfad5dc.png"/>
</details>
<details>
<summary>Docker registry with basic auth selected:</summary>
<img src="https://user-images.githubusercontent.com/139310/98481326-c627c300-21f9-11eb-8417-9c11c68f60cb.png"/>
</details>
<details>
<summary>Docker registry with Image Registry selected:</summary>
<img src="https://user-images.githubusercontent.com/139310/98481330-c7f18680-21f9-11eb-9367-dfa9f4e67bd2.png"/>
</details>
<details>
<summary>Git server auth type options:</summary>
<img src="https://user-images.githubusercontent.com/139310/98481332-ca53e080-21f9-11eb-83ad-eedd8b068ce1.png"/>
</details>
<details>
<summary>Git server with basic auth selected:</summary>
<img src="https://user-images.githubusercontent.com/139310/98481335-ccb63a80-21f9-11eb-820c-24b8286262c5.png"/>
</details>
<details>
<summary>Git server with SSH Key selected:</summary>
<img src="https://user-images.githubusercontent.com/139310/98481338-cf189480-21f9-11eb-93e5-81a5a53e0801.png"/>
</details>

<!--
![old-layout](https://user-images.githubusercontent.com/139310/98481378-1141d600-21fa-11eb-9796-9dcf45ad34c9.png)
![new-layout-default-settings](https://user-images.githubusercontent.com/139310/98481342-d5a70c00-21f9-11eb-8b97-e43024d78f7f.png)
![access-to-options](https://user-images.githubusercontent.com/139310/98481321-c1fba580-21f9-11eb-8853-8731f6b94175.png)
![docker-registry-auth-type-options](https://user-images.githubusercontent.com/139310/98481324-c45dff80-21f9-11eb-9dc2-94ba6bfad5dc.png)
![docker-registry-basic-auth](https://user-images.githubusercontent.com/139310/98481326-c627c300-21f9-11eb-8417-9c11c68f60cb.png)
![docker-registry-image-registry](https://user-images.githubusercontent.com/139310/98481330-c7f18680-21f9-11eb-9367-dfa9f4e67bd2.png)
![git-server-options](https://user-images.githubusercontent.com/139310/98481332-ca53e080-21f9-11eb-83ad-eedd8b068ce1.png)
![git-server-basic-auth](https://user-images.githubusercontent.com/139310/98481335-ccb63a80-21f9-11eb-820c-24b8286262c5.png)
![git-server-ssh-key](https://user-images.githubusercontent.com/139310/98481338-cf189480-21f9-11eb-93e5-81a5a53e0801.png)
-->

**Unit test coverage report**: 
No unit test changed.

**Test setup:**
* Install OpenShift Pipelines Operator
* Create a new pipeline (e.g. add a deployment and enable the option "Add Pipeline")
* Open the "Start Pipeline" dialog

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

jerolimov commented:
/kind bug
/cc @karthikjeeyar

karthikjeeyar commented:
@jerolimov , The Server URL annotation's value can be `https://github.com` (for: https://github.com/test/testRepo) or `github.com` (for ssh url: git@github.com/test/testRepo), It depends on the git url that we use in the pipelines.

jerolimov commented:
@karthikjeeyar You're absolute right here. We need to make a difference here between http and ssh. I updated this PR (incl the PR description and gif). Wdyt about this change now?

(I didn't tested the dockerconfig way and are not sure if https:// is correct in that case.)

karthikjeeyar commented:
@jerolimov Thank you for improving the hint based on the Authentication type, this is very useful to the user, but however i have a suggestion to reorder the `Authentication Type` and the `Annotation section`, because i feel changing the previous field's help text will not help the user to identify the value upfront, they might have to revisit the previous field to adjust the value. I think we need UX input here.
![Auth_Type_reorder](https://user-images.githubusercontent.com/9964343/97403087-c7174580-1919-11eb-806b-141809a1c396.png)

 @bgliwa01 Let us know your thought on this.

jerolimov commented:
@karthikjeeyar Yes, I see this exactly like you and had this idea as well, from the code perspective it looks a little bit strange, but I think also this would improve the UX. I reordered the components to get a better feeling for the result:

![image](https://user-images.githubusercontent.com/139310/97416437-ba3d2680-1906-11eb-97dd-e08d3e829f8f.png)
(Not part of this PR yet!)

@karthikjeeyar @bgliwa01 Wdyt? I removed the subtitle "Designate provider to be authenticated" because I was unsure where we want keep it. But I can restore it again if you want this.

jerolimov commented:
Btw, if the tekton [Authentication at Run Time](https://github.com/tektoncd/pipeline/blob/master/docs/auth.md#configuring-ssh-auth-authentication-for-git) here contains all cases we also should (in this ticket, or in a new one..) hide some combinations here:

![image](https://user-images.githubusercontent.com/139310/97437886-66d9d100-1924-11eb-9a43-f4d0224e8f5a.png)
(screenshot of the doc toc)

* When "Access to" "Git Server" is selected we should hide the "Auth Type" option "Image Registry Credentials"
* When "Access to" "Docker Server" is selected we should hide the "Auth Type" option "SSH auth"

/cc @andrewballantyne @vdemeester wdyt? Should I create a new ticket for this or solve this here as well?

christianvogt commented on frontend/packages/dev-console/src/components/pipelines/modals/common/SecretAnnotation.tsx:
```diff
@@ -1,16 +1,24 @@
 };
 
+const helpText = {
+  [SecretType.dockerconfigjson]: 'The base server url (e.g. https://quay.io/)',
+  [SecretType.basicAuth]: 'The base server url (e.g. https://github.com)',
+  [SecretType.sshAuth]: 'Server hostname without schema or path (e.g. github.com)',
```
Localize these.

jerolimov replied:
We have PR https://github.com/openshift/console/pull/7143 by @vikram-raj  for pipeline i18n. I would like to keep it there. Depending on whats merged first we should update the other PR, I'm fine to rebase this one if the other is merged first.

jerolimov replied:
Now localized

vdemeester commented:
> * When "Access to" "Git Server" is selected we should hide the "Auth Type" option "Image Registry Credentials"
> * When "Access to" "Docker Server" is selected we should hide the "Auth Type" option "SSH auth"
> 
> /cc @andrewballantyne @vdemeester wdyt? Should I create a new ticket for this or solve this here as well?

Right, I think it make sense to hide thoses

jerolimov commented:
@christianvogt @vdemeester @karthikjeeyar @andrewballantyne Updated the PR with your feedback,

* it nows also reorders the fields so that you need to select the "Access to" and "Auth type" fields before entering the "Server URL".
* it hides the "Auth type" "Image Registry Credentials" for Git servers and "SSH auth" for Docker registry.

Btw, should we rename Docker registry to Image registry?!?

jerolimov commented:
@bgliwa01 Thanks for bringing this up in the weekly pipeline sync. Added the following changes now:

* Changed section title from "Create Source Secret" to "Create Secret"
* Restored "Designate provider to be authenticated" as hint for "Access to:"
* Renamed "Docker Registry" to "Image Registry"

![updated-pr](https://user-images.githubusercontent.com/139310/98698833-c5686b80-2376-11eb-9f65-e68a1a8a7ffb.png)

jerolimov commented:
/retest

karthikjeeyar commented:
Thanks @jerolimov
/lgtm

jerolimov commented:
/retest

bgliwa01 reviewed (approved):
lgtm

andrewballantyne commented:
Perhaps the fields should be emptied (reset to defaults) when you switch the various controls. Anything that is under the control should return to the defaults. I can set a Git URL, and go "oh no wait, I have an SSH key instead!" and switching that will not clear the "https://github.com" url, and thus subtlety not work. I, for one, would not look at a hint for a field that I already entered a value for.

andrewballantyne commented:
This will need to be rebased on the new pipeline-plugin package. Hopefully it's straight forward as the folder structure should be relatively untouched.

bgliwa01 commented:
> Perhaps the fields should be emptied (reset to defaults) when you switch the various controls. Anything that is under the control should return to the defaults. I can set a Git URL, and go "oh no wait, I have an SSH key instead!" and switching that will not clear the "https://github.com" url, and thus subtlety not work. I, for one, would not look at a hint for a field that I already entered a value for.

+1

jerolimov commented:
Rebased to the new file structure of the pipeline-plugin

jerolimov commented:
/retest

jerolimov commented:
/retest

karthikjeeyar commented:
@jerolimov  needs rebase again :(

jerolimov commented:
Translation does not contain the components under public/components/secrets and looks now like this:

![image](https://user-images.githubusercontent.com/139310/99520349-4802b380-2993-11eb-8a96-86646c0fbacd.png)

jerolimov commented:
/retest

vikram-raj commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/SecretForm.tsx:
```diff
@@ -48,20 +42,73 @@ const renderSecretForm = (
 
+  const secretTypes = React.useMemo<Record<string, string>>(
+    () => ({
+      [SecretAnnotationId.Git]: t('pipelines-Git Server'),
+      [SecretAnnotationId.Image]: t('pipelines-Image Registry'),
```
```suggestion
      [SecretAnnotationId.Git]: t('pipelines-plugin~pipelines-Git Server'),
      [SecretAnnotationId.Image]: t('pipelines-plugin~pipelines-Image Registry'),
```

jerolimov replied:
Thanks :+1:

vikram-raj commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/SecretForm.tsx:
```diff
@@ -48,20 +42,73 @@ const renderSecretForm = (
+  // Uses a memo instead of const outside of the function so that we can add i18n right here
+  const helpText = React.useMemo(
+    () => ({
+      [SecretType.dockerconfigjson]: t('pipelines-The base server url (e.g. https://quay.io/)'),
+      [SecretType.basicAuth]: t('pipelines-The base server url (e.g. https://github.com)'),
+      [SecretType.sshAuth]: t('pipelines-Server hostname without schema or path (e.g. github.com)'),
```
```suggestion
      [SecretType.dockerconfigjson]: t('pipelines-plugin~The base server url (e.g. https://quay.io/)'),
      [SecretType.basicAuth]: t('pipelines-plugin~The base server url (e.g. https://github.com)'),
      [SecretType.sshAuth]: t('pipelines-plugin~Server hostname without schema or path (e.g. github.com)'),
```

jerolimov replied:
Thanks :+1:

jerolimov commented:
/retest

jerolimov commented:
Thanks @vikram-raj, I added the missing translation namespace.

jerolimov commented:
/retest

jerolimov commented:
/retest

vikram-raj reviewed (commented):
verified it. Thanks @jerolimov 

/lgtm

jerolimov commented:
> Forgot to hit a review on Friday. @jerolimov You still have the same issue that I noted in my description to Brigid. If you switch the type or access-to it does not reset the value of the Server URL... which could be easily overlooked that the hint changed.

So you want that we clear the server URL or change them as well? Yeah its maybe the safest way.

We could also apply some additional form validation rules here. But let us start with the easiest solution (resetting it)?! Will do that.

andrewballantyne commented:
> > Forgot to hit a review on Friday. @jerolimov You still have the same issue that I noted in my description to Brigid. If you switch the type or access-to it does not reset the value of the Server URL... which could be easily overlooked that the hint changed.
> 
> So you want that we clear the server URL or change them as well? Yeah its maybe the safest way.
> 
> We could also apply some additional form validation rules here. But let us start with the easiest solution (resetting it)?! Will do that.

Right, the easiest thing is for us to clear (reset back to default) items below the changed field... that way the user is mentally required to re-address those values. I think you'll probably also want to reset the field back to untouched(?) because Formik won't understand what we are doing and will flag the field against the validation schema -- which says "string not empty"... and errors it because we just emptied it haha. I don't think this noise is needed for the user.

jerolimov commented:
@andrewballantyne @vikram-raj @bgliwa01 I added the following change so that the "Server URL" was now cleared when

* the user switches the "Access to" and this automatically switch the the "Authentication Type"
* the user manually switches "Authentication Type"

![image](https://user-images.githubusercontent.com/139310/99962657-de1e4b80-2d90-11eb-9732-a12e2329a2aa.png)

jerolimov commented:
/retest

jerolimov commented:
> JavaScript heap out of memory

/retest

jerolimov commented:
/test backend

karthikjeeyar commented:
Thanks @jerolimov, Verified locally, works fine
/lgtm
